File: frontend/src/store/zoteroNoteSync.ts
// Per-paper state for the "Create Note from Annotations" Zotero sync. Persisted
// server-side (per-user, per-paper) so an enabled paper resumes syncing when
// the user reopens it (and so the ZoteroPanel reflects the last sync result
// even after a reload). `syncing` is ephemeral — never persisted.

import { create } from "zustand";
import * as api from "../lib/api";

export interface PaperNoteSync {
  enabled: boolean;
  /** Cached Zotero key of the annotations child note (lets the backend PATCH
   *  directly instead of listing children each sync). */
  noteKey: string | null;
  /** Cached Zotero key of the parent paper item (avoids re-searching each sync). */
  parentKey: string | null;
  /** ms epoch of the last SUCCESSFUL sync (not advanced on error). */
  lastSyncedAt: number | null;
  lastError: string | null;
  /** annotations included in the last successful sync. */
  lastCount: number;
  syncing: boolean;
  /** content signature of the last-written note HTML (skip-if-unchanged). */
  contentSig: string | null;
}

export const DEFAULT_PAPER_SYNC: PaperNoteSync = {
  enabled: false,
  noteKey: null,
  parentKey: null,
  lastSyncedAt: null,
  lastError: null,
  lastCount: 0,
  syncing: false,
  contentSig: null,
};

interface ZoteroNoteSyncState {
  papers: Record<string, PaperNoteSync>;
  loaded: boolean;
  load: () => Promise<void>;
  reset: () => void;
  setEnabled: (arxivId: string, enabled: boolean) => void;
  beginSync: (arxivId: string) => void;
  finishSync: (
    arxivId: string,
    r: {
      noteKey?: string;
      parentKey?: string;
      count: number;
      error?: string;
      /** drop cached noteKey+parentKey so the next run rediscovers (used when
       *  an upsert failed, e.g. the note/parent was deleted in Zotero). */
      clearKeys?: boolean;
      contentSig?: string;
    }
  ) => void;
}

export const useZoteroNoteSyncStore = create<ZoteroNoteSyncState>((set) => ({
  papers: {},
  loaded: false,

  load: async () => {
    try {
      const remote = await api.listNoteSync();
      const papers: Record<string, PaperNoteSync> = {};
      for (const [arxivId, n] of Object.entries(remote)) {
        papers[arxivId] = {
          enabled: n.enabled,
          noteKey: n.noteKey,
          parentKey: n.parentKey,
          lastSyncedAt: n.lastSyncedAt,
          lastError: n.lastError,
          lastCount: n.lastCount,
          syncing: false, // ephemeral
          contentSig: n.contentSig,
        };
      }
      set({ papers, loaded: true });
    } catch {
      set({ loaded: true });
    }
  },

  reset: () => set({ papers: {}, loaded: false }),

  setEnabled: (arxivId, enabled) =>
    set((s) => {
      const prev = s.papers[arxivId] || DEFAULT_PAPER_SYNC;
      const next: PaperNoteSync = {
        ...prev,
        enabled,
        // clearing the error on enable so a stale failure doesn't linger;
        // on disable we also clear it so the panel stops showing errors.
        lastError: null,
      };
      void api.putNoteSync(arxivId, {
        enabled, lastError: null,
      }).catch(() => { /* non-fatal */ });
      return { papers: { ...s.papers, [arxivId]: next } };
    }),

  beginSync: (arxivId) =>
    set((s) => ({
      papers: {
        ...s.papers,
        [arxivId]: { ...(s.papers[arxivId] || DEFAULT_PAPER_SYNC), syncing: true },
      },
    })),

  finishSync: (arxivId, r) =>
    set((s) => {
      const prev = s.papers[arxivId] || DEFAULT_PAPER_SYNC;
      const noteKey = r.clearKeys ? null : r.noteKey ?? prev.noteKey ?? null;
      const parentKey = r.clearKeys ? null : r.parentKey ?? prev.parentKey ?? null;
      const next: PaperNoteSync = {
        ...prev,
        syncing: false,
        lastSyncedAt: r.error ? prev.lastSyncedAt : Date.now(),
        lastError: r.error ?? null,
        lastCount: r.count,
        noteKey,
        parentKey,
        contentSig: r.contentSig ?? prev.contentSig,
      };
      void api.putNoteSync(arxivId, {
        noteKey, parentKey,
        lastSyncedAt: next.lastSyncedAt,
        lastError: next.lastError,
        lastCount: next.lastCount,
        contentSig: next.contentSig,
      }).catch(() => { /* non-fatal */ });
      return { papers: { ...s.papers, [arxivId]: next } };
    }),
}));

File: backend/app/routers/zotero_note_sync.py
"""Zotero note-sync state router — per-user, per-paper sync state.

Tracks the cached noteKey/parentKey, last sync result, and enabled flag for the
"Create Note from Annotations" Zotero flow. `syncing` is ephemeral and never
persisted (no sync is in flight across a reload).
"""
from __future__ import annotations

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlmodel import select
from sqlmodel.ext.asyncio.session import AsyncSession

from ..db import get_session
from ..deps import current_user
from ..models import User, ZoteroNoteSyncRow

router = APIRouter(prefix="/zotero-note-sync", tags=["zotero-note-sync"])


class NoteSyncOut(BaseModel):
    enabled: bool
    noteKey: str | None
    parentKey: str | None
    lastSyncedAt: int | None
    lastError: str | None
    lastCount: int
    contentSig: str | None


class NoteSyncPatch(BaseModel):
    enabled: bool | None = None
    noteKey: str | None = None
    parentKey: str | None = None
    lastSyncedAt: int | None = None
    lastError: str | None = None
    lastCount: int | None = None
    contentSig: str | None = None


def _to_out(row: ZoteroNoteSyncRow) -> NoteSyncOut:
    return NoteSyncOut(
        enabled=row.enabled, noteKey=row.note_key, parentKey=row.parent_key,
        lastSyncedAt=row.last_synced_at, lastError=row.last_error,
        lastCount=row.last_count, contentSig=row.content_sig,
    )


@router.get("", response_model=dict[str, NoteSyncOut])
async def list_note_sync(
    user: User = Depends(current_user),
    session: AsyncSession = Depends(get_session),
) -> dict[str, NoteSyncOut]:
    rows = (
        await session.exec(
            select(ZoteroNoteSyncRow).where(ZoteroNoteSyncRow.user_id == user.id)
        )
    ).all()
    return {r.arxiv_id: _to_out(r) for r in rows}


@router.put("/{arxiv_id}", response_model=NoteSyncOut)
async def put_note_sync(
    arxiv_id: str,
    body: NoteSyncPatch,
    user: User = Depends(current_user),
    session: AsyncSession = Depends(get_session),
) -> NoteSyncOut:
    row = await session.get(ZoteroNoteSyncRow, (user.id, arxiv_id))
    if row is None:
        row = ZoteroNoteSyncRow(
            user_id=user.id, arxiv_id=arxiv_id,
            enabled=body.enabled if body.enabled is not None else True,
            note_key=body.noteKey, parent_key=body.parentKey,
            last_synced_at=body.lastSyncedAt, last_error=body.lastError,
            last_count=body.lastCount or 0, content_sig=body.contentSig,
        )
        session.add(row)
    else:
        if body.enabled is not None:
            row.enabled = body.enabled
        if body.noteKey is not None:
            row.note_key = body.noteKey
        if body.parentKey is not None:
            row.parent_key = body.parentKey
        if body.lastSyncedAt is not None:
            row.last_synced_at = body.lastSyncedAt
        if body.lastError is not None:
            row.last_error = body.lastError
        if body.lastCount is not None:
            row.last_count = body.lastCount
        if body.contentSig is not None:
            row.content_sig = body.contentSig
        session.add(row)
    await session.commit()
    await session.refresh(row)
    return _to_out(row)
